### PR TITLE
SONARJAVA-2519 properly enqueue exception handlers

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 1]
```diff
@@ -555,12 +710,25 @@ private boolean handleMethodInvocation(Instruction instruction) {
...
+              ).forEach(ps -> {
+                ps.storeExitValue();
+                programPosition.block.successors().stream()
+                  .map(b -> (BytecodeCFGBuilder.Block) b)
+                  .filter(b -> b.getExceptionType() != null)
+                  .filter(b -> b.getExceptionType().charAt(0) == '!'
+                    ||/*required as long as there is no real type tracking*/ exceptionType == null
+                    || exceptionType.isSubtypeOf(b.getExceptionType()))
+                  .forEach(b -> {
+                      System.out.println("foo");
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MAJOR Make this line start at column 56. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)
MAJOR Replace this use of System.out or System.err by a logger. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS106)

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 2]
```diff
@@ -555,12 +710,25 @@ private boolean handleMethodInvocation(Instruction instruction) {
...
+                    enqueue(new ProgramPoint(b), ps);}
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MAJOR Move this closing curly brace to the next line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [BytecodeEGWalker.java#L254](https://github.com/SonarSource/sonar-java/blob/1e98f87fe993f6758fa8a0ab83348221d1ab4c10/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L254): Remove this unused "exitBlock" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)
1. ![MINOR][MINOR] [BytecodeEGWalker.java#L254](https://github.com/SonarSource/sonar-java/blob/1e98f87fe993f6758fa8a0ab83348221d1ab4c10/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L254): Remove the "exitBlock" field and declare it as a local variable in the relevant methods. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1450)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: java-frontend/src/main/java/org/sonar/java/bytecode/cfg/BytecodeCFGBuilder.java [hunk 3]
```diff
@@ -114,12 +114,21 @@ public boolean isVarArgs() {
...
+      return blocks;
+    }
   }
   public static class Block implements CFG.IBlock<Instruction> {
     int id;
     BytecodeCFG cfg;
     List<Instruction> instructions;
     List<Block> successors;
+
+    public String getExceptionType() {
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/bytecode/cfg/BytecodeCFGBuilder.java]: we don't have any rules about it, but I think it will be better to keep fields together and methods separated

File: java-frontend/src/test/java/org/sonar/java/bytecode/se/BytecodeEGWalkerExecuteTest.java [hunk 4]
```diff
@@ -926,4 +932,63 @@ private static ProgramState setDoubleOrLong(ProgramState programState, SymbolicV
       return programState.removeConstraintsOnDomain(sv, BytecodeEGWalker.StackValueCategoryConstraint.class);
     }
   }
+
+  @Test
+  public void test_enqueuing_only_happy_path() {
```
Comment (saberduck) [inline java-frontend/src/test/java/org/sonar/java/bytecode/se/BytecodeEGWalkerExecuteTest.java]: this test file is already quite big and these tests have different focus and structure, I think it would be better to have them in a new file

File: java-frontend/src/test/java/org/sonar/java/bytecode/se/BytecodeEGWalkerExecuteTest.java [hunk 5]
```diff
@@ -422,7 +428,7 @@ public void test_dup2_x2_form4() throws Exception {
   @Test
   public void test_add_sub_mul_div_rem() throws Exception {
     int[] opcodes = new int[] {
-      Opcodes.IADD, Opcodes.LADD, Opcodes.FADD, Opcodes.DADD,
+      Opcodes.IADD, LADD, Opcodes.FADD, Opcodes.DADD,
```
Comment (saberduck) [inline java-frontend/src/test/java/org/sonar/java/bytecode/se/BytecodeEGWalkerExecuteTest.java]: unnecessary change

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 6]
```diff
@@ -555,12 +705,21 @@ private boolean handleMethodInvocation(Instruction instruction) {
...
+          .forEach(yield -> {
+            org.sonar.plugins.java.api.semantic.Type exceptionType = yield.exceptionType();
             yield.statesAfterInvocation(
-              Lists.reverse(pop.values), Collections.emptyList(), pop.state, () -> constraintManager.createExceptionalSymbolicValue(yield.exceptionType())).forEach(ps -> {
-              ps.storeExitValue();
-              enqueue(new ProgramPoint(exitBlock), ps);
-            }));
+              Lists.reverse(pop.values), Collections.emptyList(), pop.state, () -> constraintManager.createExceptionalSymbolicValue(exceptionType)
+              ).forEach(ps -> {
+                ps.storeExitValue();
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: I don't see a test where this is used (should be an invocation throwing an exception which is not caught )

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 7]
```diff
@@ -555,12 +705,21 @@ private boolean handleMethodInvocation(Instruction instruction) {
...
+                programPosition.block.successors().stream()
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: I think this deserves to be extracted to the method, it's quite hard to follow, e.g. `List<Block> exceptionHandlers(...)`

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 8]
```diff
@@ -555,12 +705,21 @@ private boolean handleMethodInvocation(Instruction instruction) {
...
+                  .map(b -> (BytecodeCFGBuilder.Block) b)
+                  .filter(b -> b.getExceptionType() != null)
+                  .filter(b -> b.getExceptionType().charAt(0) == '!'
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: this condition ought to be named properly via variable or small method

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 9]
```diff
@@ -635,8 +794,10 @@ void handleBlockExit(ProgramPoint programPosition) {
       pair.a.stream().forEach(s -> enqueue(falsePP, s));
       pair.b.stream().forEach(s -> enqueue(truePP, s));
     } else {
-      // TODO : filter some node of the EG depending of the exceptionType in the successor.
-      programPosition.block.successors().forEach(b -> enqueue(new ProgramPoint(b), programState));
+      programPosition.block.successors().stream()
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: also this pipeline could be extracted to aptly named method imo

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java (hunk 6 above)
Comment (benzonico) [reply java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: assertion added.

File: java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java [hunk 10]
```diff
@@ -555,12 +706,15 @@ private boolean handleMethodInvocation(Instruction instruction) {
...
+          .forEach(yield -> {
+            org.sonar.plugins.java.api.semantic.Type exceptionType = yield.exceptionType();
             yield.statesAfterInvocation(
-              Lists.reverse(pop.values), Collections.emptyList(), pop.state, () -> constraintManager.createExceptionalSymbolicValue(yield.exceptionType())).forEach(ps -> {
-              ps.storeExitValue();
-              enqueue(new ProgramPoint(exitBlock), ps);
-            }));
+              Lists.reverse(pop.values), Collections.emptyList(), pop.state, () -> constraintManager.createExceptionalSymbolicValue(exceptionType)
+              ).forEach(ps -> {
+                ps.storeExitValue();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java]: MAJOR Make this line start at column 15. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AIndentationCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
